artemis-protocols/artemis-mqtt-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/mqtt/MQTTProtocolManagerFactory.java

```diff
@@ -64,4 +74,46 @@ public ProtocolManager createProtocolManager(ActiveMQServer server,
    public String getModuleName() {
       return MODULE_NAME;
    }
+
+
+   @Override
+   public void loadProtocolServices(ActiveMQServer server, List<ActiveMQComponent> services) {
+      services.add(new MQTTPeriodicTasks(server.getScheduledPool()));
+   }
+
+   public class MQTTPeriodicTasks implements ActiveMQComponent {
+      ScheduledExecutorService pool;
+      Runnable runnable;
+
+      public MQTTPeriodicTasks(ScheduledExecutorService pool) {
+         this.pool = pool;
+      }
+
+      @Override
+      public void start() throws Exception {
+         runnable = () -> {
+            for (Map.Entry<String, MQTTSessionState> entry : sessionStates.entrySet()) {
+               MQTTSessionState state = entry.getValue();
+               MQTTLogger.LOGGER.debug("Inspecting session state: " + state);
+               if (!state.getAttached() && defaultMqttSessionExiryInterval != -1 && state.getDisconnectedTime() + (defaultMqttSessionExiryInterval * 1000) < System.currentTimeMillis()) {
+                  MQTTLogger.LOGGER.debug("Removing expired session state: " + state);
+                  sessionStates.remove(entry.getKey());
+               }
+            }
+         };
+         pool.scheduleAtFixedRate(runnable, 0, 5, TimeUnit.SECONDS);
+      }
+
+      @Override
+      public void stop() throws Exception {
+         if (pool instanceof ThreadPoolExecutor) {
+            ((ThreadPoolExecutor)pool).remove(runnable);
```

Comment:
the best would been to cancel on the future...

but if you extend the ScheduledComponent.. it would been easier to manage this.